### PR TITLE
Fix axis range issue in matplotlib contour plot

### LINKED ISSUE #3197: Fix the axes of `optuna.visualization.matplotlib.plot_contour`
I found that the axes of `optuna.visualization.matplotlib.plot_contour` has bug.

```python
import optuna
from optuna.testing.visualization import prepare_study_with_trials
import matplotlib.pyplot as plt

study = prepare_study_with_trials(more_than_three=True, n_objectives=2, with_c_d=True)
print(study.trials_dataframe())
figure = optuna.visualization.plot_contour(study, target=lambda t: t.values[0])

figure.write_image('plotly.png')
optuna.visualization.matplotlib.plot_contour(study, target=lambda t: t.values[0])
plt.savefig('matplotlib.png')
```

plotly

matplotlib

To plot a-b subplot, the max and the min of params are determined with trials whose params have `a` and `b`. But the maximum and minimum values should be determined independently for `a` and `b`.

_Originally posted by @not522 in https://github.com/optuna/optuna/issues/2959#issuecomment-997638266_

### PR DESCRIPTION
**Pull request by harupy**
## Motivation

Fix #3197

## Description of the changes

When computing the subplot ranges for parameter X, ignore other parameters.

## Test

Run this script and confirm the subplot ranges are correct:

```python
import optuna
from optuna.testing.visualization import prepare_study_with_trials
import matplotlib.pyplot as plt

study = prepare_study_with_trials(more_than_three=True, n_objectives=2, with_c_d=True)
print(study.trials_dataframe())
figure = optuna.visualization.plot_contour(study, target=lambda t: t.values[0])

figure.write_image('plotly.png')
optuna.visualization.matplotlib.plot_contour(study, target=lambda t: t.values[0])
plt.savefig('matplotlib.png')
```

### Before

image

### After

image

### CODE REVIEW

File: optuna/visualization/matplotlib/_contour.py [hunk 1]
```diff
@@ -226,8 +234,18 @@ def _calculate_griddata(
     x_values = []
     y_values = []
     z_values = []
+    x_range_values = []
+    y_range_values = []
     for trial in trials:
-        if x_param not in trial.params or y_param not in trial.params:
```
Comment (himkt) [inline optuna/visualization/matplotlib/_contour.py:230]: > To plot a-b subplot, the max and the min of params are determined with trials whose params have a and b. But the maximum and minimum values should be determined independently for a and b.

:+1:

File: optuna/visualization/matplotlib/_contour.py [hunk 2]
```diff
@@ -226,8 +234,18 @@ def _calculate_griddata(
...
+        contains_x_parm = x_param in trial.params
+        if contains_x_parm:
+            x_range_values.append(trial.params[x_param])
+
+        contains_y_param = y_param in trial.params
+        if contains_y_param:
+            y_range_values.append(trial.params[y_param])
+
+        if not contains_x_parm or not contains_y_param:
```
Comment (himkt) [inline optuna/visualization/matplotlib/_contour.py:248]:
```suggestion
        contains_x_param = x_param in trial.params
        if contains_x_param:
            x_range_values.append(trial.params[x_param])

        contains_y_param = y_param in trial.params
        if contains_y_param:
            y_range_values.append(trial.params[y_param])

        if not contains_x_param or not contains_y_param:
```

**Review by himkt — COMMENTED** [review]
Thank you so much for fixing @harupy! LGTM with one minor suggestion.

File: optuna/visualization/matplotlib/_contour.py (hunk 2 above)
Comment (harupy) [reply optuna/visualization/matplotlib/_contour.py:248]: Thanks for the catch!

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! The change looks good to me. By the way, do you have any idea to test this change in unit-tests?

**Comment by harupy** [comment]
I think we could do something like this:

```python
subplots = fig.get_axes()  # not sure if this really works

# If we search params a, b, and c, `subplots` looks like:
# a [[subplot1, subplot2, subplot3],
# b  [subplot4, subplot4, subplot6],
# c  [subplot7, subplot8, subplot9]],
#     a         b         c

# check subplot1, 2, and 3 have the same y-axis range that covers the searched space for a.
assert all(ax.get_ylim() <= (a_min, a_max) for ax in subplots[0, :])

# subplot4, 5, and 6 should have the same y-axis range that covers the searched space for b.
assert all(ax.get_ylim() <= (b_min, b_max) for ax in subplots[1, :])

...
```

**Comment by HideakiImamura** [comment]
Looks reasonable. Could you add the test in this PR?

**Comment by harupy** [comment]
@HideakiImamura @himkt Added a test!

File: tests/visualization_tests/matplotlib_tests/test_contour.py [hunk 3]
```diff
@@ -235,3 +238,53 @@ def test_generate_contour_plot_for_few_observations(params: List[str]) -> None:
...
+
+def range_covers(range1: Tuple[float, float], range2: Tuple[float, float]) -> bool:
+    """Returns True if `range1` covers `range2`"""
+    min1, max1 = sorted(range1)
+    min2, max2 = sorted(range2)
+    return min1 <= min2 and max1 >= max2
+
+
+def test_contour_subplots_have_correct_axis_labels_and_ranges() -> None:
+    study_without_trials = prepare_study_with_trials()
```
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:256]: [question] Why did you name `study_without_trials` for the returned value of `prepare_study_with_trials` here? :eyes:

File: tests/visualization_tests/matplotlib_tests/test_contour.py [hunk 4]
```diff
@@ -235,3 +238,53 @@ def test_generate_contour_plot_for_few_observations(params: List[str]) -> None:
     study = prepare_study_with_trials(less_than_two=True)
     figure = plot_contour(study, params)
     assert not figure.has_data()
+
+
+def all_equal(iterable: Iterable) -> bool:
+    """Returns True if all the elements are equal to each other"""
+    return len(set(iterable)) <= 1
```
Comment (himkt) [inline tests/visualization_tests/matplotlib_tests/test_contour.py:245]:
```suggestion
    return len(set(iterable)) == 1
```

I think `len(set(iterable))` returning `0` is not the expected scenario so it looks safer to check if the value is exactly one. Please let me know if you have the intention I missed.

**Review by himkt — COMMENTED** [review]
Thank you so much @harupy for adding the test as well! Looks solid with few questions.

**Review by HideakiImamura — APPROVED** [review]
Thanks for adding tests. LGTM.

File: tests/visualization_tests/matplotlib_tests/test_contour.py (hunk 4 above)
Comment (harupy) [reply tests/visualization_tests/matplotlib_tests/test_contour.py:245]: Agree!

File: tests/visualization_tests/matplotlib_tests/test_contour.py [hunk 5]
```diff
@@ -235,3 +238,53 @@ def test_generate_contour_plot_for_few_observations(params: List[str]) -> None:
...
+    # - The y-axis label of subplot 7 is "param_c"
+    # - ...
+    param_ranges = {
+        "param_a": (0.0, 3.0),
+        "param_b": (0.0, 3.0),
+        "param_c": (2.0, 5.0),
+    }
+    for index, (param_name, param_range) in enumerate(param_ranges.items()):
+        minimum, maximum = param_range
+        padding = (maximum - minimum) * AXES_PADDING_RATIO
```
[the remark this answers is not in the document]
Comment (harupy) [reply tests/visualization_tests/matplotlib_tests/test_contour.py:281]: Do I need to insert a comment indicating this?

File: tests/visualization_tests/matplotlib_tests/test_contour.py (hunk 3 above)
Comment (harupy) [reply tests/visualization_tests/matplotlib_tests/test_contour.py:256]: Ah I copied `study_without_trials = prepare_study_with_trials(no_trials=True)`, and then removed `no_trials=True`. I'll fix.

File: tests/visualization_tests/matplotlib_tests/test_contour.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (himkt) [reply tests/visualization_tests/matplotlib_tests/test_contour.py:281]: I'm sorry that I missed your response. :bow: I think no. The current change is enough for this patch!

**Comment by himkt** [comment]
Retrying codecov... :loading:

**Review by himkt — APPROVED** [review]
LGTM. Thank you for the contribution @harupy!

**Pull request merged**